File: cpp/src/hangman.cpp
#include "hangman.hpp"

/** Using namespaces **/
using std::cin;
using std::cout;
using std::endl;
using std::flush;
using std::string;
using std::stringstream;

/** Public methods **/
bool Hangman::startGame()
{
    bool result = playGame(10);

    if(result) 
    {
        cout << "Congratulations, you guessed the word \"" << _currentWord << "\" correctly." <<  endl;
        return true;
    }
    else 
    {
        cout << "You did not guess \"" << _currentWord << "\" correctly, and have been hanged." << endl;
        return false;
    }
}

/** Constructor **/
Hangman::Hangman(string listPath)
{
    _wordList = new WordList(listPath);
}

/** Destructor **/
Hangman::~Hangman()
{
    delete _wordList;
}

/** Private methods **/
bool Hangman::playGame(int triesLeft) 
{
    //Gets a new word from the word list and convert it to underscores
    _currentWord = _wordList->getRandomWord();
    _currentGuess = string(_currentWord.length(), '_');
    printStatus(triesLeft);

    //Asks for at new guess as long as the word has not being guessed or the user have tried enough times
    while (triesLeft > 0) 
    {
       char guess = getGuess();

       if(!isCharInWord(guess))
           triesLeft--;

       printStatus(triesLeft);
  
       if(_currentWord == _currentGuess)
           return true;
    }
    return false;
}

char Hangman::getGuess() 
{
    string input;
    
    do
    {
        cout << "Enter your next guess: " << flush;
        getline(cin, input);
    }
    while (!isalpha(input[0]));

    //Just a couple of newlines for better formating of the output
    cout << endl << endl;
    return input[0];
}

bool Hangman::isCharInWord(char guess)
{
    bool answer = false;
    char input = tolower(guess);
    unsigned int index, stringLength;
        
    //Checks if the given chars has already been already guessed 
    if(!string::npos == _guessedChars.str().find(guess))
        return true;

    //We need the length of the word so we can map correct guesses from the real word to the underscored version
    stringLength = _currentWord.length();

    //Checks if the word contains instances of the guessed guess, uses a counted loop as index is needed
    for(index =0; index < stringLength; index++)
    {
        //The guess is already converted to its lower case, so we more easily can check if it have been used
        if(_currentWord.at(index) == input || _currentWord.at(index) == toupper(input))
        {
            answer = true;
            _currentGuess.at(index) = _currentWord.at(index);
        }
    }

    //Adds the guessed character to the guessed characters string
    _guessedChars << input << " ";

    return answer;
}

void Hangman::printStatus(unsigned int triesLeft)
{
    cout << "Word: " << _currentGuess << endl;
    cout << "Tries Left: " << triesLeft << endl;
    cout << "Guessed Letters: " << _guessedChars.str() << endl << endl;

    cout << "The Gallow" << endl << endl;

    //Prints a man getting hanged depending on how many faults the player have
    switch(triesLeft) {
        case 10:
            cout << "\n\n\n\n\n\n   " << endl;
            break;
        case 9:
            cout << "\n\n\n\n\n   \n_______________\n" << endl;
            break;
        case 8:
            cout << "\n\n\n  |\n  |\n  |\n________________\n" << endl;
            break;
        case 7:
            cout << "  |\n  |\n  |\n  |\n  |\n  |\n________________\n" << endl;
            break;
        case 6:
            cout << "  |--         \n  |\n  |\n  |\n  |\n  |\n  |\n________________\n"  << endl;
            break;
        case 5:
            cout << "  |------         \n  |\n  |\n  |\n  |\n  |\n  |\n________________\n" << endl;
            break;
        case 4:
            cout << "  |------         \n  |     |\n  |\n  |\n  |\n  |\n  |\n________________\n" << endl;
            break;
        case 3:
            cout << "  |------         \n  |     |\n  |     0\n  |\n  |\n  |\n  |\n________________\n" << endl;
            break;
        case 2:
            cout << "  |------         \n  |     |\n  |     0\n  |    /|\\\n  |\n  |\n  |\n________________\n" << endl;
            break;
        case 1:
            cout << "  |------         \n  |     |\n  |     0\n  |    /|\\\n  |     |\n  |\n  |\n________________\n" << endl;
            break;
        case 0:
            cout << "  |------         \n  |     |\n  |     0\n  |    /|\\\n  |     |\n  |    / \\\n  |\n________________\n" << endl;
            break;
    }
}

File: objective-c/src/hangman.m
#import "hangman.h"

@implementation Hangman 

    /** Instance variables **/
    WordList* wordList;
    NSString* currentWord;
    NSMutableString *currentGuess;
    NSMutableString *guessedChars;

    /** Constructor **/
    - (id)initWithWordListPath: (NSString*) listPath {
        self = [super init];
        wordList = [[WordList alloc] initWithWordListPath: listPath];
        currentWord = [wordList getRandomWord];
        return self;
    }

    /** Destructor **/
    - (void)dealloc {
        [super dealloc];
        [currentWord release];
        [currentGuess release];
        [guessedChars release];
    }

    /** Public methods **/
    - (bool)startGame  {
        bool result = [self playGame: 10];

        if(result) 
            printf("Congratulations, you guessed the word \"%s\" correctly.\n", [currentWord UTF8String]);
        else 
            printf("You did not guess \"%s\" correctly, and have been hanged.\n", [currentWord UTF8String]);

        return result;
    }

    /** Private methods **/
    - (bool)playGame: (int)triesLeft {
        currentGuess = [[NSMutableString alloc] initWithCharacter: '_' length: [currentWord length]];
        guessedChars = [[NSMutableString alloc] init];

        //We print the status for the user before the game begins so they can see how many chars the word contains
        [self printStatus: triesLeft guess: [currentGuess UTF8String] chars: [guessedChars UTF8String]];

        //The main game loop that takes guesses and decreaments tries if the user guesss wrong
        while(triesLeft > 0) {
            char guess = [self getGuess];
            //Checks if the char is in the word the user is guessing and updates guessed chars
            if(![self isCharInWord: guess])
                triesLeft--;

            //Objective-C prefers enums instead of integer as retun value so converting to an int might beconme a problem
            if([currentWord localizedCaseInsensitiveCompare: currentGuess] == NSOrderedSame)
                return true;

            [self printStatus: triesLeft guess: [currentGuess UTF8String] chars: [guessedChars UTF8String]];
        }
        return false;
    }

    - (char)getGuess {
        char tempInput = 0;
        char userInput = 0;

        while (!isalpha(userInput)) {
            printf("Enter your next guess: ");
            tempInput = getchar();
            userInput = tempInput;

            //Flushes the input buffer until a new line
            while(tempInput != '\n')
                tempInput = getchar();
        }
        return userInput;
    }

    - (bool)isCharInWord: (char) guess {
        bool correctGuess = false;
        char lowerGuess = tolower(guess);

        //Checks if the guessed char has already been tried by the user
        int guessedCharsLength = [guessedChars length];
        for(int index = 0; index < guessedCharsLength; index++)
            if([guessedChars characterAtIndex: index] == lowerGuess)
                return true;

        //The guess has not been done before, so we add it to the list
        [guessedChars appendFormat: @" %c", lowerGuess];

        //Checks if the word we are guessing contains instances of the guessed char
        int currentWordLength = [currentWord length];
        NSString* lowerGuessString = [[NSString alloc] initWithFormat: @"%c", lowerGuess];

        for(int index = 0; index < currentWordLength; index++) {
            if([currentWord characterAtIndex: index] == lowerGuess) {
                [currentGuess replaceCharactersInRange: NSMakeRange(index,1) withString: lowerGuessString];
                correctGuess = true;
            }
        }
        [lowerGuessString release];
        return correctGuess;
    }

    - (void)printStatus: (unsigned int) guessesLeft guess: (const char*) currentGuess chars: (const char*) guessedChars {
           printf("Word: %s\n", currentGuess);
           printf("Tries Left: %d\n", guessesLeft);
           printf("Guessed Letters: %s\n\n", guessedChars);

           printf("The Gallow\n\n");

           switch(guessesLeft) {
               case 10:
                   printf("\n\n\n\n\n\n\n");
                   break;
               case 9:
                   printf("\n\n\n\n\n   \n_______________\n\n");
                   break;
               case 8:
                   printf("\n\n\n  |\n  |\n  |\n________________\n\n");
                   break;
               case 7:
                   printf("  |\n  |\n  |\n  |\n  |\n  |\n________________\n\n");
                   break;
               case 6:
                   printf("  |--         \n  |\n  |\n  |\n  |\n  |\n  |\n________________\n\n");
                   break;
               case 5:
                   printf("  |------         \n  |\n  |\n  |\n  |\n  |\n  |\n________________\n\n");
                   break;
               case 4:
                   printf("  |------         \n  |     |\n  |\n  |\n  |\n  |\n  |\n________________\n\n");
                   break;
               case 3:
                   printf("  |------         \n  |     |\n  |     0\n  |\n  |\n  |\n  |\n________________\n\n");
                   break;
               case 2:
                   printf("  |------         \n  |     |\n  |     0\n  |    /|\\\n  |\n  |\n  |\n________________\n\n");
                   break;
               case 1:
                   printf("  |------         \n  |     |\n  |     0\n  |    /|\\\n  |     |\n  |\n  |\n________________\n\n");
                   break;
               case 0:
                   printf("  |------         \n  |     |\n  |     0\n  |    /|\\\n  |     |\n  |    / \\\n  |\n________________\n\n");
                   break;
           }
    }
@end
